snapshots/windows/common.go

```diff
@@ -0,0 +1,343 @@
+// +build windows
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package windows
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"path/filepath"
+	"strings"
+
+	winfs "github.com/Microsoft/go-winio/pkg/fs"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/containerd/containerd/snapshots/storage"
+	"github.com/containerd/continuity/fs"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	rootfsSizeLabel           = "containerd.io/snapshot/io.microsoft.container.storage.rootfs.size-gb"
+	reuseScratchLabel         = "containerd.io/snapshot/io.microsoft.container.storage.reuse-scratch"
+	reuseScratchOwnerKeyLabel = "containerd.io/snapshot/io.microsoft.owner.key"
+	// labelScratchSnapshotLocation is a label provided in snapshotter opts to specify
+	// if the scratch snapshot should be stored in a different location specified by
+	// this annotations. (Only supported on windows & lcow snapshotter as of now)
+	labelScratchSnapshotLocation = "containerd.io/snapshot/io.microsoft.override-scratch"
+)
+
+// windowsSnapshotterBase is the base snapshotter for both LCOW & WCOW snapshotters. It provides common
+// methods required for both snapshotters.
+type windowsSnapshotterBase struct {
+	root     string
+	ms       *storage.MetaStore
+	snConfig *WindowsSnapshotterConfig
+}
+
+// WindowsSnapshotterConfig is the configuration related to windows snapshotters (i.e LCOW & WCOW)
+type WindowsSnapshotterConfig struct {
```

Comment:
Dumb question, where is this able to be edited by a client? Can you supply this on the ctrd config in some way?

Reply:
Asking as I see that here it's created with default values https://github.com/kevpar/containerd/pull/27/files#diff-aa911c498121f3429af552dce390cdd14a9200b4d283e9f2dcefe205d8248223R30

Reply:
Yeah, it's not very intuitive. For the snapshotter plugin named "windows", if we have a section named `[plugins.windows]` in the toml, during containerd startup, it will be automatically decoded (into the struct provided with that plugin) here: https://github.com/kevpar/containerd/blob/master/services/server/server.go#L119.